### DISCUSSION TITLE
Correcting cost at a geofence

### DISCUSSION THREAD

**Discussion by srivaths** [teslamate-org/teslamate#2409] [Q&A]
The cost of charging at one of my geofences had not been set.  When I set it I was asked if I wanted to retroactively set the cost of all charges at that location (several hundred instances).  I agreed since that is what I wanted.

It was only then that I noticed that I had a typo in the cost I had specified.  However when you change the charging cost at a geofence that previously had a cost value you are not prompted for retroactive changes.

Is there a way to easily apply a cost change retroactively?

(I am aware that I can go into the Charges dashboard and apply it for each charging event.  However there are about 500 instances and this would be very time consuming).

**Comment by cwanja** [comment]
Could you delete the geofence, set it back up, then reset the charging cost?

**Reply by srivaths** [reply]
@cwanja Thanks for the tip. I was thinking of that too.  I am trying to figure out if there would be any other unintended side effects of this approach.  The geofence in question is my home so it shows up **a lot**! 🙂

**Reply by cwanja** [reply]
I am not sure if it will actually clear the costs though now that I think about it. It would just remove the label from the location. I do not think it clears the cost associated with it. A good test could be trial it on a single location and see what happens. 

The other way to do it would be to drop the costs from the database by geofence. But this would clear **all** costs, regardless of if the cost is right or wrong. 

As always. Backup your data before regardless.

**Reply by srivaths** [reply]
@cwanja You are right.  Removing a geofence does not clear the costs.  I tested on another geofence.
Do you know how to drop the cost from the database?  Although comfortable with database queries I haven't messed with Teslamate at the code level.  I am hoping there is a way to `Update SOME_TABLE set cost=null where location=SOME_GEOFENCE`

**Reply by cwanja** [reply]
First and foremost, **backup your data**. Not going to warranty this at all. 

So within the database, there is a table called `charging_processes` that has `geofence_id` and `cost`. You can run `update charging_processes set cost = NULL where geofence_id = '1';`

Where geofence_id is the ID of the one you want to clear out. Note this will clear **all** charging costs for that geofence. So make sure it is what you want to do. You can pull the `geofence_id` from TeslaMate URL when you are editing the geofence:
`http://URL:4000/geo-fences/1/edit`

**Reply by srivaths** [reply]
@cwanja Thank you very much for pointing me to the table and the field name.  I need to brush up on my Docker foo to connect to the proper container and do the necessary.  I'll work on it.  Thanks, again.

**Reply by cwanja** [reply]
> @cwanja Thank you very much for pointing me to the table and the field name. I need to brush up on my Docker foo to connect to the proper container and do the necessary. I'll work on it. Thanks, again.

Sure thing. One other thing you could filter on would be by `modified date`. That might narrow the charges to only those that have the wrong calculated costs. You could always do a `select` first to make sure you are looking at the right rows before running the `update` command.

**Reply by srivaths** [reply]
Oh absolutely.  Always `select` first to validate before doing such bulk `update`s 😃

**Reply by srivaths** [reply]
@cwanja Paydirt!

0. Backup the data
1. `update charging_processes set cost = NULL where geofence_id = '1'`
2. Pull up Geofence and set the correct cost/kWh.  Press Save.
3. You are prompted asking whether the cost should be applied retroactively.  Accept
4. Pull up **Charges** dashboard.
5. Yay!!

**Reply by cwanja** [reply]
Awesome, glad it worked!

**Reply by hadfiiw** [reply]
Sorry to bring this topic back up but does anybody have the steps to connect to the proper Docker container?  For us Docker-impaired individuals.  :-P

**Reply by cwanja** [reply]
@hadfiiw - you should be able to use `docker-compose exec database psql teslamate teslamate`. The password (`teslamate`) might be different and `database` could be `db` depending on your YML file.

See more here https://docs.teslamate.org/docs/maintenance/manually_fixing_data

**Reply by hadfiiw** [reply]
Thanks @cwanja that worked perfectly!

For reference, here is what I ran...

```
sudo docker-compose exec database psql teslamate teslamate
```
which changed to a "teslamate=#" prompt.  Then (using geofence 4 per the geofence's URL)...

```
teslamate=# select from charging_processes where geofence_id = '4';
--
(88 rows)

teslamate=# update charging_processes set cost = NULL where geofence_id = '4';
UPDATE 88
teslamate=# 
```

**Comment by scottrbaxter** [comment]
```
teslamate=# select from charging_processes where geofence_id = '4';
--
(88 rows)

teslamate=# update charging_processes set cost = NULL where geofence_id = '4';
UPDATE 88
teslamate=#
```

this worked perfectly for me to retroactively set charging costs - i initially entered a decimal point incorrectly. thanks!

curious, year over year we've had a change in electricity pricing. i'd like to go back and set all records for 2023 and 2024 to be the value charged per kwh on our utility bill. since we can't set the grafana `Cost/kWh` value directly, could someone point me to the best way to set this value for earlier dates in the postres db?

**Reply by cwanja** [reply]
'Ol ChatGPT to the rescue. 

```
SELECT *
FROM your_table
WHERE your_date_column BETWEEN '2023-01-01' AND CURRENT_DATE;
```
